Client/src/ChronologClientImpl.cpp

```diff
@@ -83,91 +150,142 @@ int chronolog::ChronologClientImpl::CreateChronicle(std::string const& chronicle
                                      const std::unordered_map<std::string, std::string> &attrs,
                                      int &flags) 
 {  
+    if( chronicle_name.empty() )
+    {  return CL_ERR_INVALID_ARG; }
+
     std::lock_guard<std::mutex> lock_client(chronologClientMutex);
 
     if((clientState == UNKNOWN) || (clientState == SHUTTING_DOWN))
     {  return CL_ERR_NO_CONNECTION; }
 
-    return rpcClient_->CreateChronicle(chronicle_name, attrs, flags);
+    return rpcVisorClient->CreateChronicle(chronicle_name, attrs, flags);
 }
 
 //TODO: client account must be passed into the rpc call 
 int chronolog::ChronologClientImpl::DestroyChronicle(std::string const& chronicle_name) 
 {
+    if( chronicle_name.empty())
+    {  return CL_ERR_INVALID_ARG; }
+
     std::lock_guard<std::mutex> lock_client(chronologClientMutex);
 
     if((clientState == UNKNOWN) || (clientState == SHUTTING_DOWN))
     {  return CL_ERR_NO_CONNECTION; }
 
-    return rpcClient_->DestroyChronicle(chronicle_name);
+    return rpcVisorClient->DestroyChronicle(chronicle_name);
 }
 
 //TODO: client account must be passed into the rpc call 
 int chronolog::ChronologClientImpl::DestroyStory(std::string const& chronicle_name, std::string const& story_name)
 {
+    if(chronicle_name.empty() || story_name.empty())
+    {  return CL_ERR_INVALID_ARG; }
+
     std::lock_guard<std::mutex> lock_client(chronologClientMutex);
 
     if((clientState == UNKNOWN) || (clientState == SHUTTING_DOWN))
     {  return CL_ERR_NO_CONNECTION; }
 
-    return rpcClient_->DestroyStory(chronicle_name, story_name);
+    return rpcVisorClient->DestroyStory(chronicle_name, story_name);
 }
 
 std::pair<int,chronolog::StoryHandle*> chronolog::ChronologClientImpl::AcquireStory(std::string const& chronicle_name, std::string const& story_name,
                                   const std::unordered_map<std::string, std::string> &attrs, int &flags) 
 {
+    std::cout <<"ChronologClientImpl::AcquireStory call : "<< chronicle_name <<":"<< story_name<<std::endl;
+    if(chronicle_name.empty() || story_name.empty())
+    {  return std::pair<int, chronolog::StoryHandle*>(CL_ERR_INVALID_ARG, nullptr); }
+
     std::lock_guard<std::mutex> lock_client(chronologClientMutex);
 
     if((clientState == UNKNOWN) || (clientState == SHUTTING_DOWN))
     {  return std::pair<int, chronolog::StoryHandle*>(CL_ERR_NO_CONNECTION,nullptr); }
 
-    chronolog::StoryHandle * storyHandle = nullptr;
-    //TODO: this function should return StoryId & vector<KeeperIdCard>  that will be used to create StoryWritingHandle
+    // this function can be called from any client thread, so before sending an rpc request to the Visor
+    // we check if the story acquisition request has been already granted to the process on some other thread
+    // 
+    chronolog::StoryHandle * storyHandle = storyteller->findStoryWritingHandle(chronicle_name,story_name);
```

Comment:
@fkengun Connect can indeed be called from any thread and that's why it's guarded by the mutex, making sure that only connect rpc request to the Visor is issued at any point in time from the same ChronologClientImpl singleton object.  ChronologClientImpl::AcquireStory would not proceed beyond line#201 if there's no valid connection already established by the client application process.

Reply:
@ibrodkin I got how the singleton ChronologClientImpl works now. line 94 and 122 make sure there is only on StorytellerClient in the single ChronologClientImpl object.
reusing local StorytellerClient without checking with Visor on the latest metadata whether a new AcquireStory() should be accepted could still be a problem. but we can come back to this later since we do not have any API to change the permission. currently in DestroyStory(), the only API for now that may affect this, it already returns error when the Story is in acquisition.